src/extent.c

```diff
@@ -363,6 +363,43 @@ extents_remove_locked(tsdn_t *tsdn, extents_t *extents, extent_t *extent,
 	    cur_extents_npages - (size >> LG_PAGE), ATOMIC_RELAXED);
 }
 
+/*
+ * Find an extent with size [min_size, max_size) to satisfy the alignment
+ * requirement.  For each size, try only the first extent in the heap.
+ */
+static extent_t *
+extents_fit_alignment(extents_t *extents, size_t min_size, size_t max_size,
+    size_t alignment) {
+        pszind_t pind = sz_psz2ind(extent_size_quantize_ceil(min_size));
+        pszind_t pind_max = sz_psz2ind(extent_size_quantize_ceil(max_size));
+
+	for (pszind_t i = (pszind_t)bitmap_ffu(extents->bitmap,
+	    &extents_bitmap_info, (size_t)pind); i < pind_max; i =
+	    (pszind_t)bitmap_ffu(extents->bitmap, &extents_bitmap_info,
+	    (size_t)i+1)) {
+		assert(i < NPSIZES);
+		assert(!extent_heap_empty(&extents->heaps[i]));
+		extent_t *extent = extent_heap_first(&extents->heaps[i]);
+		assert(extent_size_get(extent) >= min_size);
+		uintptr_t base = (uintptr_t)extent_base_get(extent);
+		size_t esize = extent_size_get(extent);
+
+		uintptr_t next_align = ALIGNMENT_CEILING((uintptr_t)base,
```

Comment:
It looks like we don't do any overflow checking -- right?

Reply:
Right. Doesn't seem to be needed (if this overflows, we won't and can't use the current extent).

Reply:
Can you add a comment with some math? I'm definitely not following.

Suppose there's a very large alignment. Then next_align will be less then base, so we'll pass the first check, so leadsize will be very big, so esize - leadsize will be very big, so we'll return an improper extent, no?

Reply:
Will add comments. However if the next_align is less than base, we will skip this extent in the first check, right?

Reply:
Hold on. That wasn’t right. Let me fix this.